code/game/gamemodes/setupgame.dm

```diff
@@ -120,7 +120,7 @@
 		if(G.block)
 			if(G.block in blocks_assigned)
 				warning("DNA2: Gene [G.name] trying to use already-assigned block [G.block] (used by [english_list(blocks_assigned[G.block])])")
-			dna_genes.Add(G)
+			dna_genes[G.type] = G

```

Comment:
The only thing you cannot (easily) use as key in associative lists are integers, as BYOND will attempt to add/update that particular index instead of using the hash value.